acceptance/helpers.rb

```diff
@@ -0,0 +1,130 @@
+# @return [Hash] An options hash to pass to BeakerPuppet's `install_puppet_agent_on` method
+def agent_install_options
+  agent_version = ENV['FROM_AGENT_VERSION'] || ENV['PUPPET_CLIENT_VERSION'] # This is the legacy name from puppet 3 / module 1.x
+
+  if agent_version
+    unless dev_builds_accessible?
+      # The user requested a specific build, but they can't download from internal sources
+      env_var_name = ENV['FROM_AGENT_VERSION'] ? 'FROM_AGENT_VERSION' : 'PUPPET_CLIENT_VERSION'
+      fail_test(<<-WHY
+  You requested a specific build of puppet-agent, but you don't have access to
+  Puppet's internal build servers. You can either:
+
+  - Unset the #{env_var_name} environment variable to accept the latest Puppet 4
+    agent release (this is the defualt), or
+  - Set the $FROM_PUPPET_COLLECTION environment variable to 'puppet5' or 'puppet6' to
+    use the latest releases from those streams.
+
+  WHY
+      )
+    end
+
+    return { puppet_agent_version: agent_version }
+  end
+
+  { puppet_collection: (ENV['FROM_PUPPET_COLLECTION'] || 'pc1').downcase }
+end
+
+# Installs the puppet_agent module on the target host, and then uses the PMT to install puppet_agent's dependencies
+def install_modules_on(host)
+  install_dev_puppet_module_on(host, {
+    source: File.join(File.dirname(__FILE__), '..', ),
+    module_name: 'puppet_agent'
+  })
+
+  on(host, puppet('module', 'install', 'puppetlabs-stdlib',     '--version', '4.16.0'), { acceptable_exit_codes: [0] })
+  on(host, puppet('module', 'install', 'puppetlabs-inifile',    '--version', '2.1.0'),  { acceptable_exit_codes: [0] })
+  on(host, puppet('module', 'install', 'puppetlabs-apt',        '--version', '4.4.0'),  { acceptable_exit_codes: [0] })
+  on(host, puppet('module', 'install', 'puppetlabs-transition', '--version', '0.1.1'),  { acceptable_exit_codes: [0] })
+end
+
+# Pass these options to any `with_puppet_running_on` block
+def master_puppetconf
+  {
+      master: { autosign: true, dns_alt_names: master }
+  }
+end
+
+# These options are used to configure puppet-agent
+def agent_puppetconf
+  {
+      agent: { # TODO: change to main?
+          server: "#{master}",
+          ssldir: "$vardir/ssl"  # Necessary?
+      }
+  }
+end
+
+def run_setup(use_master: false)
+  logger.notify("Setup: Install puppet-agent on agents")
+
+  # Install the puppet-agent package and configure it
+  agents_only.each do |agent|
+    install_puppet_agent_on(agent, agent_install_options)
+    configure_puppet_on(agent, agent_puppetconf)
+  end
+
+  # Either intsall modules or clear SSL/firewall, based on whether this is a masterless run
+  if use_master
+    logger.notify("Setup [master/agent test]: Clean SSL and disable firewall on all hosts")
+    # If using a master, make sure no certificate weirdness happens when the new agent contacts it:
+    hosts.each do |host|
+      on(host, "rm -rf '#{host.puppet['ssldir']}'")
+      stop_firewall_with_puppet_on(host)
```

Comment:
Could this be done in the pre-suite? I see no code in the teardown that sets the firewall back on for these hosts.

Reply:
I guess? This variant of the method (with use_master set to true) hasn't been exercised yet and I'm not even sure it would work in its current state.

The stop firewall method requires puppet to be installed first -- if I did that in the pre-suite, I'd need to install puppet-agent on the agents in the pre-suite in addition to on the master, which doesn't make much sense to me. I could write a beaker-puppet helper that didn't rely on puppet to stop firewalls, but that seems like a lot of heavy lifting for this initial chunk of work.

Reply:
Yeah that makes sense. It would be useful to add a comment about why we're stopping the firewall there instead of in the pre-suite.